File: memory-api/src/main/java/com/memory/config/aop/ControllerLoggingAspect.java
package com.memory.config.aop;

import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.*;
import org.springframework.stereotype.Component;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import jakarta.servlet.http.HttpServletRequest;
import java.util.Arrays;

@Slf4j
@Aspect
@Component
public class ControllerLoggingAspect {

    @Before("execution(* com.memory.controller..*(..))")
    public void logControllerEntry(JoinPoint joinPoint) {
        HttpServletRequest request = getHttpServletRequest();
        if (request != null) {
            String method = request.getMethod();
            String url = request.getRequestURL().toString();
            String clientIp = getClientIp(request);
            
            log.info("🎯 [CONTROLLER] {} {} - Class: {}, Method: {}, IP: {}", 
                method, url, 
                joinPoint.getTarget().getClass().getSimpleName(),
                joinPoint.getSignature().getName(),
                clientIp
            );
            
            // 파라미터 로그 (민감정보 제외)
            Object[] args = joinPoint.getArgs();
            if (args.length > 0) {
                log.info("📝 [PARAMS] {}", Arrays.toString(args));
            }
        }
    }

    @Around("execution(* com.memory.controller..*(..))")
    public Object logExecutionTime(ProceedingJoinPoint joinPoint) throws Throwable {
        long startTime = System.currentTimeMillis();
        
        try {
            Object result = joinPoint.proceed();
            long endTime = System.currentTimeMillis();
            
            log.info("⏱️ [TIMING] {}.{} executed in {}ms", 
                joinPoint.getTarget().getClass().getSimpleName(),
                joinPoint.getSignature().getName(),
                (endTime - startTime)
            );
            
            return result;
        } catch (Exception e) {
            long endTime = System.currentTimeMillis();
            log.error("❌ [ERROR] {}.{} failed after {}ms - Error: {}", 
                joinPoint.getTarget().getClass().getSimpleName(),
                joinPoint.getSignature().getName(),
                (endTime - startTime),
                e.getMessage()
            );
            throw e;
        }
    }

    @AfterReturning(pointcut = "execution(* com.memory.controller..*(..))", returning = "result")
    public void logControllerReturn(JoinPoint joinPoint, Object result) {
        log.info("✅ [SUCCESS] {}.{} completed successfully", 
            joinPoint.getTarget().getClass().getSimpleName(),
            joinPoint.getSignature().getName()
        );
    }

    private HttpServletRequest getHttpServletRequest() {
        try {
            ServletRequestAttributes attributes = 
                (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
            return attributes.getRequest();
        } catch (Exception e) {
            return null;
        }
    }

    private String getClientIp(HttpServletRequest request) {
        String xForwardedFor = request.getHeader("X-Forwarded-For");
        if (xForwardedFor != null && !xForwardedFor.isEmpty()) {
            return xForwardedFor.split(",")[0].trim();
        }
        
        String xRealIp = request.getHeader("X-Real-IP");
        if (xRealIp != null && !xRealIp.isEmpty()) {
            return xRealIp;
        }
        
        return request.getRemoteAddr();
    }
}

File: memory-api/src/main/kotlin/com/memory/config/aop/ControllerLoggingAspect.kt
package com.memory.config.aop

import jakarta.servlet.http.HttpServletRequest
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes

@Aspect
@Component
class ControllerLoggingAspect {
    private val log = LoggerFactory.getLogger(ControllerLoggingAspect::class.java)

    @Before("execution(* com.memory.controller..*(..))")
    fun logControllerEntry(joinPoint: JoinPoint) {
        val request = getHttpServletRequest()
        if (request != null) {
            val method = request.method
            val url = request.requestURL.toString()
            val clientIp = getClientIp(request)

            log.info(
                "🎯 [CONTROLLER] {} {} - Class: {}, Method: {}, IP: {}",
                method, url,
                joinPoint.target.javaClass.simpleName,
                joinPoint.signature.name,
                clientIp
            )

            val args = joinPoint.args
            if (args.isNotEmpty()) {
                log.info("📝 [PARAMS] {}", args.contentToString())
            }
        }
    }

    @Around("execution(* com.memory.controller..*(..))")
    fun logExecutionTime(joinPoint: ProceedingJoinPoint): Any? {
        val startTime = System.currentTimeMillis()

        return try {
            val result = joinPoint.proceed()
            val endTime = System.currentTimeMillis()

            log.info(
                "⏱️ [TIMING] {}.{} executed in {}ms",
                joinPoint.target.javaClass.simpleName,
                joinPoint.signature.name,
                endTime - startTime
            )

            result
        } catch (e: Exception) {
            val endTime = System.currentTimeMillis()
            log.error(
                "❌ [ERROR] {}.{} failed after {}ms - Error: {}",
                joinPoint.target.javaClass.simpleName,
                joinPoint.signature.name,
                endTime - startTime,
                e.message
            )
            throw e
        }
    }

    @AfterReturning(pointcut = "execution(* com.memory.controller..*(..))", returning = "result")
    fun logControllerReturn(joinPoint: JoinPoint, result: Any?) {
        log.info(
            "✅ [SUCCESS] {}.{} completed successfully",
            joinPoint.target.javaClass.simpleName,
            joinPoint.signature.name
        )
    }

    private fun getHttpServletRequest(): HttpServletRequest? {
        return try {
            val attributes = RequestContextHolder.currentRequestAttributes() as ServletRequestAttributes
            attributes.request
        } catch (e: Exception) {
            null
        }
    }

    private fun getClientIp(request: HttpServletRequest): String {
        val xForwardedFor = request.getHeader("X-Forwarded-For")
        if (!xForwardedFor.isNullOrEmpty()) {
            return xForwardedFor.split(",")[0].trim()
        }

        val xRealIp = request.getHeader("X-Real-IP")
        if (!xRealIp.isNullOrEmpty()) {
            return xRealIp
        }

        return request.remoteAddr
    }
}
